nixos/modules/services/web-apps/mattermost.nix

```diff
@@ -166,179 +207,126 @@ in
-          '';
-        };
+        description = lib.mdDoc "Addtional configuration options as Nix attribute set in config.json schema.";
       };
     };
   };
 
-  config = mkMerge [
-    (mkIf cfg.enable {
-      users.users = optionalAttrs (cfg.user == "mattermost") {
-        mattermost = {
-          group = cfg.group;
-          uid = config.ids.uids.mattermost;
-          home = cfg.statePath;
-        };
-      };
+  config = mkIf cfg.enable {
+    users.users.${cfg.user} = {
+      home = cfg.dataDir;
+      group = cfg.group;
+      isSystemUser = true;
+    };
 
-      users.groups = optionalAttrs (cfg.group == "mattermost") {
-        mattermost.gid = config.ids.gids.mattermost;
-      };
+    users.groups.${cfg.group} = { };
+
+    services.postgresql = {
+      enable = true;
+      ensureDatabases = [ cfg.database.name ];
+      ensureUsers = [{
+        name = cfg.user;
+        ensurePermissions."DATABASE ${cfg.database.name}" = "ALL PRIVILEGES";
+      }];
+    };
 
-      services.postgresql.enable = cfg.localDatabaseCreate;
+    systemd.tmpfiles.rules = [
+      "d ${cfg.dataDir} 0750 ${cfg.user} ${cfg.group} - -"
+      "d ${cfg.logDir} 0750 ${cfg.user} ${cfg.group} - -"
+      "d ${cfg.configDir} 0750 ${cfg.user} ${cfg.group} - -"
+      "d ${mutableDataDir} 0750 ${cfg.user} ${cfg.group} - -"
+
+      # Remove and recreate tempDir.
+      "R ${tempDir} - - - - -"
+      "d ${tempDir} 0750 ${cfg.user} ${cfg.group} - -"
+
+      # Ensure that pluginDir is a directory, as it could be a symlink on prior versions.
+      "r ${pluginDir} - - - - -"
+      "d ${pluginDir} 0750 ${cfg.user} ${cfg.group} - -"
+      "d ${mattermostConf.PluginSettings.Directory} 0750 ${cfg.user} ${cfg.group} - -"
+      "d ${mattermostConf.PluginSettings.ClientDirectory} 0750 ${cfg.user} ${cfg.group} - -"
+
+      "L+ ${cfg.dataDir}/fonts - - - - ${cfg.package}/fonts"
+      "L+ ${cfg.dataDir}/i18n - - - - ${cfg.package}/i18n"
+      "L+ ${cfg.dataDir}/templates - - - - ${cfg.package}/templates"
+      "L+ ${cfg.dataDir}/client - - - - ${cfg.package}/client"
+    ] ++ (
+      if mattermostPlugins == null then
+        # Create the plugin tarball directory if it's a symlink.
+        [
+          "r ${cfg.dataDir}/plugins - - - - -"
+          "d ${cfg.dataDir}/plugins 0750 ${cfg.user} ${cfg.group} - -"
+        ]
+      else
+        # Symlink the plugin tarball directory.
+        ["L+ ${cfg.dataDir}/plugins - - - - ${mattermostPlugins}"]
+    );
 
-      # The systemd service will fail to execute the preStart hook
-      # if the WorkingDirectory does not exist
-      system.activationScripts.mattermost = ''
-        mkdir -p "${cfg.statePath}"
+    systemd.services.mattermost = {
+      description = "Mattermost chat service";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" "postgresql.service" ];
+      requires = [ "network.target" "postgresql.service" ];
```

Comment:
There should be an option to run the postgres on another machine or not use it at all.